Feature: Config API and logger

Verdanix opened the pull request:
# Feature Pull Request

## Feature Description
A logging utility class and configuration model API and manager were introduced.  

One logging class removes the need for @Slf4j to be used everywhere, leading to longer startup times. Using a single logger reduces resource consumption and ensures an easy way to swap out loggers if required.   
- The config model API was developed to support the features we only needed. Gestalt is great, but it had many unneeded and unwanted features that weren't valuable to the Nora project. Additionally, using only 2 classes reduces resource consumption and startup time. 
- Created a GH action to run maven tests 10 times to ensure the tests aren't fluked

## Implementation Details
### Logger
NoraLogger is a single thread-safe class that allows swapping between loggers if the need arises easily.

It uses the Bill Pugh Singleton design pattern, which ensures "built-in" thread safety without using the synchronized keyword.

### Configuration API
The Configuration API utilizes 2 main classes: the Config and ConfigManager class. 

**Config Class**
The Config class has 2 methods for importing and exporting. It uses the Java Properties class as a simplified interface for importation and exportation.

**ConfigManager**
Secure and robust measures include checking whether a file is a regular, writable, and not a symbolic link. A while loop runs on a separate daemon thread for auto file reloading to ensure it doesn't affect the main thread. It uses Java's WatchService API for easy reloading. 

**Conclusion**
Both classes utilize the synchronized keyword,  Lombok's NonNull annotation, and the final keyword to ensure collaborators know what should and shouldn't be reassigned.
 
## API Changes
N/A

## Checklist
- [x] Code follows the project's style guidelines
- [x] All new and existing tests pass
- [x] New unit tests were added for the feature
- [x] Integration tests cover the new functionality
- [x] Documentation was updated to reflect the new feature

Verdanix committed:
Initial commit

- Started work on the core module

-Setup logback logger

Verdanix committed:
Added a static logger utility class.

Verdanix committed:
Added unit tests for NoraLogger

- Added normal log tests

Added formatted log messgae tests

- Added throwable log tests

- Added file output tests

Verdanix committed:
Updated testng.xml

- Added parallel tests

Verdanix committed:
Updated NoraLoggerTest

- Added parameter null tests to NoraLoggerTest

Verdanix committed:
Removed gestalt and added exception + unit tests

- Removed gestalt. It didn't live up to the hype, and had features Nora didn't need

- Added a custom exception for configurations

Added unit tests for the exception

Verdanix committed:
Added mockito testng support

Verdanix committed:
Updated javadocs

Added html unordered lists instead of just dashes

Verdanix committed:
Updated javadoc comments

- Replaced the dashes with unordered HTML lists

- Added package-info.java

Verdanix committed:
Created properties  manager

- Added a model-based config system

- Added a ConfigManager to create, save, and reload properties files

Verdanix committed:
Changed NoraLogger to final class

Verdanix committed:
Updated testng.xml and test classes

- Removed parallel processing for the ConfigManager test

- Combined the setup add cleanup  methods into 1

Verdanix committed:
Created GH Action for code testing

Verdanix committed:
Updated ConfigManager and GH workflow

- Added a file creation if the file doesn't exist

- Code testing now runs 10 times to ensure it's not 'fluked'

trammel-may commented on .github/workflows/code_testing_action.yml:
```diff
@@ -0,0 +1,28 @@
+          mvn clean install
+      - name: Test
+        run: |
+          mvn test
+          mvn test
+          mvn test
+          mvn test
+          mvn test
+          mvn test
+          mvn test
+          mvn test
+          mvn test
+          mvn test
```
why so many test executions?

Verdanix replied:
To ensure a single test run isn't a fluke. During development for other projects, I've had them fail about 20% of the time


Going that extra mile is just to add a bit of test safety

trammel-may commented on core/src/main/java/dev/masonroot/nora/NoraLogger.java:
```diff
@@ -0,0 +1,325 @@
+ *       classpath.
+ * </ul>
+ */
+@Slf4j
+public final class NoraLogger {
+  private NoraLogger() {}
```
This class is unnecessary

Verdanix replied:
The class reduces startup time because the `@Slf4j` annotation takes around 400MS on that single annotation

Having one class is just better than way

trammel-may commented on core/src/test/java/dev/masonroot/nora/NoraLoggerTest.java:
```diff
@@ -0,0 +1,251 @@
+  @SneakyThrows
+  @AfterSuite(alwaysRun = true)
+  public void testSetup() {
+    Files.delete(this.filePath);
```
static import Files

Verdanix replied:
For all the static imports

https://stackoverflow.com/questions/420791/what-is-a-good-use-case-for-static-import-of-methods

Verdanix replied:
For all the static imports

https://stackoverflow.com/questions/420791/what-is-a-good-use-case-for-static-import-of-methods

trammel-may replied:
What are you referencing on this page?

trammel-may commented on core/src/test/java/dev/masonroot/nora/NoraLoggerTest.java:
```diff
@@ -0,0 +1,251 @@
+
+  @BeforeTest(alwaysRun = true)
+  public void testLogFileExists() {
+    Assert.assertTrue(Files.exists(this.filePath));
```
static insert Assert

trammel-may commented on core/src/test/java/dev/masonroot/nora/NoraLoggerTest.java:
```diff
@@ -0,0 +1,251 @@
+      groups = {"info"})
+  public void testInfoLog() {
+    String message = "This is an info log";
+    NoraLogger.info(message);
```
static import NoraLogger

trammel-may commented on core/src/test/java/dev/masonroot/nora/NoraLoggerTest.java:
```diff
@@ -0,0 +1,251 @@
+  public void testInfoLog() {
+    String message = "This is an info log";
+    NoraLogger.info(message);
+    AssertJUnit.assertTrue(Files.readString(this.filePath).contains(message));
```
static import AssertJUnit

trammel-may commented on core/src/test/java/dev/masonroot/nora/NoraLoggerTest.java:
```diff
@@ -0,0 +1,251 @@
+        Files.readString(this.filePath).contains(message + "formatted info log"));
+  }
+
+  @SneakyThrows
```
why is this being sneaky thrown? Generally this is a no no and should be avoided.

Verdanix replied:
Only for the Files.readString

trammel-may commented on core/src/test/java/dev/masonroot/nora/NoraLoggerTest.java:
```diff
@@ -0,0 +1,251 @@
+import org.testng.annotations.BeforeTest;
+import org.testng.annotations.Test;
+
+public class NoraLoggerTest {
```
A couple of things

1. Avoid writing tests which read or write from the file system. 
2. I wouldn't bother writing a test for this class because its a thin wrapper to begin with
3. The tests are essentially verifying the behaviors of the logging library underneath and not your class, so these tests are not unit tests
4. The NoraLogger does not need to exist

trammel-may commented on core/src/test/java/dev/masonroot/nora/config/ConfigManagerTest.java:
```diff
@@ -0,0 +1,148 @@
+  @BeforeTest(alwaysRun = true)
+  @AfterTest(alwaysRun = true)
+  public void beforeAndAfterFileDeletion() {
+    Files.deleteIfExists(this.filePath);
```
Same as above, do not write tests which depend on the file system

trammel-may commented on core/src/test/java/dev/masonroot/nora/exceptions/ConfigManagerExceptionTest.java:
```diff
@@ -0,0 +1,33 @@
+public class ConfigManagerExceptionTest {
+  @Test(description = "Test ConfigManagerException")
+  public void testConfigManagerExceptionWithMessage() {
+    ConfigManagerException exception = new ConfigManagerException("Test message");
```
You should not have to write unit tests for an exception as they are not units of code, they are data. Exceptions should and will be covered by verifying exceptions are thrown at the desired times in unit tests that test business logic.

Test logic, do not test data.

trammel-may commented on core/src/main/java/dev/masonroot/nora/config/Config.java:
```diff
@@ -0,0 +1,123 @@
+ *       Implementing classes can override this method to provide custom validation logic.
+ * </ul>
+ */
+public interface Config {
```
This is not a good config interface. Your config interface should expose your configuration in a strict way. 

I have no idea what configuration is legal within this system. All I know is that the config object presumably keeps an instance of Properties which it can spit back out. Consider the following configuration interface:


```java
public interface Configuration {
   public boolean isDebug();
   public String getEnvironment();
   public int getPortNumber();
   public List<FeatureFlag> getFeatureFlags();
 }
```

This config object offers much more strict promised to what it can provide. The consumer of the configuration object should not have to know what all the magic environment variable names are, or the property names, or any of how the configuration is passed into the application. The consumer of configuration wants to have type-safe access to configuration data.

trammel-may commented on core/src/main/java/dev/masonroot/nora/config/ConfigManager.java:
```diff
@@ -0,0 +1,530 @@
+  public void create() throws ConfigManagerException {
+    if (!this.exists()) {
+      try {
+        Files.createFile(this.configPath);
```
unnecessary this

Verdanix replied:
I'd argue it reduces the amount of brain power you need when reading the code.

Sure, it may be unnecessary when there isn't a local scope variable with the same name, but it helps me understand that it's referencing a global variable. My eyes naturally look for the local variable declaration when I don't see the this keyword.

It's also super helpful, especially if IDEA's variable renaming feature decides to break and you are forced to use the find and replace feature

Funnily enough, this has happened more than a few times. I have to relaunch IDEA. It's a whole process.

trammel-may commented on core/src/main/java/dev/masonroot/nora/config/ConfigManager.java:
```diff
@@ -0,0 +1,530 @@
+  private void startFileWatcher() {
+    try {
+      synchronized (this) {
+        this.watchService = this.configPath.getParent().getFileSystem().newWatchService();
+        this.configPath
```
unnecessary this

trammel-may commented on core/src/main/java/dev/masonroot/nora/config/ConfigManager.java:
```diff
@@ -0,0 +1,530 @@
+          new ConfigManagerException("Failed to start file watcher", e);
+      NoraLogger.error(
+          "Failed to start file watcher. This only means the file won't be reloaded at runtime.",
+          exception);
```
this logs a ghost exception that is not thrown. Could just log e.

trammel-may commented on core/src/main/java/dev/masonroot/nora/config/ConfigManager.java:
```diff
@@ -0,0 +1,530 @@
+ * @param <T> the type of configuration model extending {@code Config}
+ */
+@Getter
+public final class ConfigManager<T extends Config> {
```
I don't see where the generic type is relevant. This appears to only care about the Config parent type. The child type information is not used.

trammel-may commented on core/src/main/java/dev/masonroot/nora/config/ConfigManager.java:
```diff
@@ -0,0 +1,530 @@
+import java.util.Properties;
+import lombok.Getter;
+import lombok.NonNull;
+
```
Conceptually I like the idea of a class that handles and reloads configuration. In practice I am not sure what the purpose of this is though. I doubt end-users are intended to go edit a config file so all configuration that is modified would be done through the application would which modify the in-memory representation of the configuration first which would then be saved. 

Why is there concern for on-disk configuration after it has been read?

trammel-may commented on core/src/test/java/dev/masonroot/nora/config/ConfigModel.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.Properties;
+import lombok.NonNull;
+
+public class ConfigModel implements Config {
```
This class expresses what I discussed above. It highlights how there is a desire to have these precise configuration details yet they are all being obfuscated by a properties object. 

Set those properties free

Verdanix committed:
Removed NoraLogger's unit tests

Removed logger from testng.xml

Verdanix committed:
Removed exception tests

Verdanix committed:
Removed ConfigManagerExceptionTest from testng.xml

Verdanix committed:
Updated the Configuration API contract

- Added a stricter contract for configurations

- Updated tests to accomodate changes
